services/ingestion/event_subscriber.go

```diff
@@ -169,6 +172,9 @@ func (r *RPCEventSubscriber) subscribe(ctx context.Context, height uint64) <-cha
 						continue
 					}
 				}
+				for _, evt := range blockEvents.Events {
+					r.keystore.UnlockKey(evt.TransactionID)
```

Comment:
**Note:** I am not a big fan of having to pass around the `Keystore` instance. I would rather have that instance take care the release of used keys. It could be as simple as releasing keys after a fixed time period (such as 15-20 seconds), as it doesn't take more time for a Flow transaction to seal.

Reply:
> Note: I am not a big fan of having to pass around the Keystore instance.

Agree. We could pass an interface with just `LockKey/UnlockKey` methods, so the event subscriber cannot use the key store to sign anything.

> I would rather have that instance take care the release of used keys. 

I think ideally we need both. It's not enough to check only the block events. A failed tx might not produce any event, but we should still UnlockKey. The UnlockKey here is more for the happy path, which can unlock the key ASAP for it to be reused. 

If we rely on a fixed time period, then we would have to prepare lots of keys in case GW needs to send lots of tx within that fixed time period, otherwise, we might run out of keys.

For the fixed time period, we could keep track of tx's reference height and compare with the latest sealed height, since each tx has a expire time and a reference block, if a tx has not been sealed after X blocks, then it will never be sealed. [Check this out](https://github.com/onflow/flow-go/blob/master/model/flow/constants.go#L22).

Reply:
shouldn't unlock be at transaction result ? like sign the transaction, send, check result, when result is sealed or executed ( I think executed should also work here ) then release the key.

Reply:
> shouldn't unlock be at transaction result ? like sign the transaction, send, check result, when result is sealed or executed ( I think executed should also work here ) then release the key.

@bluesign That's what we do for the happy path, when the Flow transaction is executed, and receive the `EVM.TransactionExecuted` event, then we release the key. But if, for whatever reason, the EVM transaction is invalid, then we won't get an `EVM.TransactionExecuted` event, to release the key for that Flow transaction. That's why we have a go-routine to does it on an interval. We could also try to fetch the Flow transaction result, but we might get rate-limited from the AN.

Reply:
@m-Peter thanks I didn't think about the rate limit.